docs/whitepaper/whitepaper.md

```diff
@@ -0,0 +1,591 @@
+  abstract redirect(url, status);
+  abstract pass();
+  abstract error(err);
+}
+```
+
+#### 4.3.1.1 X.509 Certificate Plugin
+
+The X.509 Certificate plugin facilitates clients authentication by allowing them to present a certificate instead of operating with authentication tokens.
+This technically allows calling clients to assume the identities of the validator nodes through the REST API without having to have access to the signing private key of said validator node.
+
+PassportJS already has plugins written for client certificate validation, but we go one step further with this plugin by providing the option to obtain CA certificates from the validator nodes themselves at runtime.
+
+### 4.3.3 Storage Plugins
+
+Storage plugins allow the higher level packages to store and retrieve configuration metadata for a `BIF` cluster such as:
+* Who are the active validators and what are the hosts where said validators are accessible over a network
+* What public keys belong to which validator nodes
+* What transactions have been scheduled, started, completed
+
+```typescript
+interface StoragePlugin {
+  async get<T>(key: string): Promise<T>;
+  async set<T>(key: string, value: T): Promise<void>;
+}
+```
+
+# 5. Identities, Authentication, Authorization
+
+`BIF` aims to provide a unified API surface for managing identities of an identity owner.
+Developers using the `BIF` REST API for their applications can support one or both of the below requirements:
+1. Applications with a focus on access control and business process efficiency (usually in the enterprise)
+2. Applications with a focus on individual privacy (usually consumer-based applications)
+
+The following sections outline the high-level features of `BIF` that make the above vision reality.
+
+An end user (through a user interface) can issue API requests to
+* register a username+password account (with optinal 2FA) **within** `BIF`.
+* associate their wallets to their `BIF` account and execute transactions involving those registered wallet (transaction signatures performed either locally or remotely as explained above).
+
+## 5.1 Transaction Signing Modes, Key Ownership
+
+An application developer using `BIF` can choose to enable users to sign their transactions locally on their user agent device without disclosing their private keys to `BIF` or remotely where `BIF` stores private keys server-side, encrypted at rest, made decryptable through authenticating with their `BIF` account.
+Each mode comes with its own pros and cons that need to be carefully considered at design time.
+
+### 5.1.1 Client-side Transaction Signing
+
+Usually a better fit for consumer-based applications where end users have higher expectation of individual privacy.
+
+**Pros**
+* Keys are not compromised when a `BIF` deployment is compromised
+* Operator of `BIF` deployment is not liable for breach of keys (same as above)
+* Reduced server-side complexity (no need to manage keys centrally)
+
+**Cons**
```

Comment:
Presumably we can get better tradeoffs (or at least offer more fine-grained choices in terms of modes) if we use threshold cryptography.  For instance, we could have a client threshold recovery process.

Reply:
Wow, didn't think of this, but I love it! When I think about it a little more, it's a great idea even on it's own as a use-case for any mobile or web application in general where users do not want to enable 2FA/MFA (privacy minded people do not like to give up their contact details just for having access to password resets).
The client side SDK could have built-in support for it in some shape or form so that end users can easily disperse their assigned key pairs to N other people. Essentially it's a multi-sig peer to peer secret vault. I'll work on adding a section with this in a follow up PR (merging this one now as we agreed during the contributors call)